runtime/src/comm/ugni/comm-ugni.c

```diff
@@ -3859,16 +3875,94 @@ inline
 void indicate_done2(int caller, rf_done_t *ack)
 {
   static rf_done_t done = 1;
-  do_remote_put(&done, caller, ack, sizeof(done), &gnr_mreg_map[caller],
-                may_proxy_false);
+  send_polling_response(&done, caller, ack, sizeof(done));
 }
 
 
 static
-inline
-void indicate_done(fork_base_info_t* b)
+void send_polling_response(void* src_addr, c_nodeid_t locale, void* tgt_addr,
+                           size_t size)
 {
-  indicate_done2(b->caller, b->rf_done);
+  gni_post_descriptor_t* post_desc;
+
+  //
+  // If we're not the polling task, send the response the regular way.
+  // Thus we only do asynchronous completion processing on polling task
+  // response transactions initiated on its firmly bound comm domain.
+  // This gets nearly all the benefit of not waiting for completions,
+  // while avoiding concurrency control entirely.
+  //
+  if (cd == NULL || !cd->firmly_bound) {
+    do_remote_put(src_addr, locale, tgt_addr, size, &gnr_mreg_map[locale],
+                  may_proxy_false);
+    return;
+  }
+
+  DBG_P_LP(DBGF_GETPUT, "SendPollResp %p -> %d:%p (%#zx)",
+           src_addr, (int) locale, tgt_addr, size);
+
+  {
+    //
+    // This pool of GNI Post descriptors is for the polling task to use
+    // for request responses (buffer releases and "done" indicators).
+    // To improve throughput, we don't wait for the remote completions
+    // when we send these.  Instead, we collect the CQ completion events
+    // later, as they come in.  We can have as many request responses in
+    // flight simultaneously as we have Post descriptors in the pool.
+    // Small-scale runs on Aries-based systems show the polling task can
+    // do about 3 minimum-cost requests ("fast" remote forks, etc.) in
+    // the minimum network latency of ~1 usec.  The pool size we use
+    // here is enough larger that it will cover the max network latency
+    // of ~1.5 usec.
```

Comment:
OOC, is 1.5 usec the max theoretical latency, or the max observed in the small-scale testing?

Reply:
See https://www.cray.com/sites/default/files/resources/CrayXCNetwork.pdf (public info), page 22.  ~1.5 usec is the measured maximum Aries network latency for 8-byte PUTs.